javaslang/src/main/java/javaslang/collection/AbstractMap.java

```diff
@@ -21,238 +21,238 @@
  *
  * @param <K> Key type
  * @param <V> Value type
- * @param <M> Map type
+ * @param <R> Map type
  * @author Ruslan Sennov
  * @since 2.0.0
  */
-abstract class AbstractMap<K, V, M extends AbstractMap<K, V, M>> implements Map<K, V> {
+abstract class AbstractMap<K, V, R extends AbstractMap<K, V, R>> implements Map<K, V> {
```

Comment:
Classical naming question - highly opinionated. I would stay with M because it is specific to AbstractMap which captures commonalities between Map impls. The M definitely denotes a Map type. Technically it is used as a result type but it does not need necessarily to. M may be (theoretically) used in arbitrary places, e.g. return type, local var etc...

If we need such param for an AbstractQueue it will be Q. That type will not be related to AbstractMap.